examples/property_prediction/pretrain_gnns/chem/pretrain_masking.py

```diff
@@ -0,0 +1,214 @@
+
+    node_feats = [
+        bg.ndata.pop('atomic_number'),
+        bg.ndata.pop('chirality_type')
+    ]
+    edge_feats = [
+        bg.edata.pop('bond_type'),
+        bg.edata.pop('bond_direction_type')
+    ]
+    # export masked nodes labels
+    masked_nodes_labels = node_feats[0][masked_nodes_indices]
+    # mask these nodes labels
+    node_feats[0][masked_nodes_indices] = 118
+
+    if args.mask_edge == 1:
+        # export masked edges labels
+        masked_edges_labels = edge_feats[0][masked_edges_indices]
+        # mask these edges labels
+        edge_feats[0][masked_edges_indices] = 5
+    else:
+        # if no edge masking
+        masked_edges_indices = 0
+        masked_edges_labels = 0
+
+    # return batched graph, node features, edge features,
+    # masked node indices, masked edge indices, masked nodes labels, masked edges labels
+    return bg, node_feats, edge_feats, \
+           (masked_nodes_indices, masked_edges_indices), (masked_nodes_labels, masked_edges_labels)
+
+
+def train(args, model_list, train_dataloader, optimizer, criterion, device):
+    if args.mask_edge == 1:
+        model, node_linear, edge_linear = model_list
+    else:
+        model, node_linear = model_list
+
+    for epoch in range(1, args.epochs):
+        model.train()
+        node_linear.train()
+        if args.mask_edge == 1:
+            edge_linear.train()
+        with tqdm.tqdm(train_dataloader) as tq_train:
+            for step, (bg, node_feats, edge_feats, masked_indices, masked_labels) in enumerate(tq_train):
+                masked_nodes_indices, masked_edges_indices = masked_indices
+                masked_nodes_labels, masked_edges_labels = masked_labels
+
+                node_feats = [n.to(device) for n in node_feats]
+                edge_feats = [e.to(device) for e in edge_feats]
+                masked_nodes_indices = torch.LongTensor(masked_nodes_indices).to(device)
+                masked_nodes_labels = torch.LongTensor(masked_nodes_labels).to(device)
+                if args.mask_edge == 1:
+                    masked_edges_indices = torch.LongTensor(masked_edges_indices).to(device)
+                    masked_edges_labels = torch.LongTensor(masked_edges_labels).to(device)
+                bg = bg.to(device)
+
+                logits = model(bg, node_feats, edge_feats)
+
+                pred_node = node_linear(logits)
+                loss_node = criterion(pred_node[masked_nodes_indices], masked_nodes_labels)
+                node_acc = compute_accuracy(pred_node[masked_nodes_indices], masked_nodes_labels)
+                if args.mask_edge == 1:
+                    pred_edge = edge_linear(logits)
+                    # for every edge, add two corresponding node feature.
+                    masked_edges_logits = pred_edge[bg.find_edges(masked_edges_indices)[0]] + \
+                                          pred_edge[bg.find_edges(masked_edges_indices)[1]]
```

Comment:
Why did you call `bg.find_edges` for two times? This doubles the cost. Also, can we directly return the edge end nodes from the collate function?

Reply:
Updated.